### DISCUSSION TITLE
EG4/Lifepower 12v/400ah battery

### DISCUSSION THREAD

**Discussion by Justice12t** [Louisvdw/dbus-serialbattery#572] [New BMS requests]
While viewing the program scan for the BMS, it gets to Lifepower and returns "incorrect data" It doesn't appear any services start as none of the devices (MTTPs, ect) go into ext control.

1. Login to root
2. run tail -f /data/log/dbus-serialbattery.ttyUSB0/current
3. As it scrolls through the list of BMS' it gets to lifepower
4. "incorrect data"
 
**Expected behavior**
Battery/BMS information showing within the Cerbo GX, connected devices (MPPTs, ect) go into ext control.

**Screenshots**
I can grab one if needed

**VenusOS (please complete the following information):**
 - Cerbo
 - 3.00 ~14

**Battery/BMS (please complete the following information):**
 - Tian Power (LifePower/Revov) assuming it's the same BMS (with instructions I could pull the battery apart and look)
 - Cells: 4
 - Interface: USB-RS485

**Additional context**
No errors during install.  Using the same cord that was supplied with the batteries that allows me to connect the battery to a PC and use the provided (by signature solar) BMS software to view the stats.  Currently the batteries are set via dips switches to #1 and #2.

**Comment by Justice12t** [comment]
Hopefully, I'm assuming correctly that my battery should work seeing the list states Tian Power (LifePower/Revov).  If I'm wrong please let me know.

**Comment by Justice12t** [comment]
Just reading through old posts.  I've got the 2 batteries connected together and then 1 cable to the Cerbo...any chance this could be the reason?

**Comment by Louisvdw** [comment]
This is your problem. Each BMS has to be connected using it's own cable.
I have not yet tested with firmware 3.00~ so that might also be an problem but I don't know.

Rather use firmware 2.92 (latest release) and connect one battery and see if that works.

**Comment by Justice12t** [comment]
I figured it was user error.  My apologizes.  I've already purchased another cable for the second battery.  I'll down grade to the latest release 2.92 and test again tonight.  Thanks for the help.

**Comment by Justice12t** [comment]
Ok back to 2.92...single battery...fresh driver install...same issue...what else am I missing. 

![error port 0](https://user-images.githubusercontent.com/121313267/211426723-7a691d55-7b7e-4d51-8fa6-9e3d3de0da1c.PNG)
![error port 1](https://user-images.githubusercontent.com/121313267/211426742-8a9d8936-c948-48a3-b7cc-7b0b7f1cf4a9.PNG)
![error 3](https://user-images.githubusercontent.com/121313267/211426752-44082a72-c46e-4b72-8565-2547dab0c241.PNG)

**Comment by Louisvdw** [comment]
The driver is installed and is running through it's tests. So that part is all working fine. 
It does not find a battery. This is in most cases a problem with the wiring or with the adapter used. So either a loose connection, or the BMS is using a different UART than the adapter you are using. 
So check that these match. There are 2 common UART types used. TTL or RS485 (sometimes also RS232)

**Comment by Justice12t** [comment]
Am I safe to assume that if this cable works connecting the battery (rs485) to a PC (usb) reading the BMS information using the provided BMS program that the cable should be good?  I've ordered a second that should be here in a few days just to double check.  I'm going to try a few other tests tonight and see if I can get it working or narrow down the challenge.

**Comment by Louisvdw** [comment]
Yes. If the cable works to a PC it should work for the driver to the GX as well

**Comment by Justice12t** [comment]
Tested the cable on both batteries with the BMS tool and both communicated without issue.  Used this cable to connect 1 battery to the Cerbo, removed all other USBs and made sure the batteries were isolated from each other.  Pushed OS v2.92 again, reinstalled the serial program.  No change.  Same output as above.  I tired both batteries and both RS485 ports on each, no change.  Could this battery(EG4 LifePower) be different than the others? Meaning possibly a different BMS than the 24v 200ah or the 48v 100ah??  I'm not sure of a good next step, anything else you'd like me to test?

**Comment by pchiquit** [comment]
I have the EG4 24v and the BMS information is displayed on the PC program. Mine says: 3QT-YS00-16SV100A-V3.5. Do you have something similar? That will give you an idea if the BMS is the same.

![image](https://user-images.githubusercontent.com/24482951/211711479-b422baa1-09b5-4c0c-9cad-83f6ec304640.png)

**Comment by Justice12t** [comment]
Looks like mine is D-4S400AH-P00-ZTR-V1.0 and BMS Version Z4SR02.  That's me reading it from a log file I had a screen shot of.  I'll connect to the batteries tomorrow and confirm this information.

**Comment by Justice12t** [comment]
![IMG_0116](https://user-images.githubusercontent.com/121313267/211825266-3424e1d0-1e4e-4916-9f3f-340a0b298936.jpg)
![IMG_0117](https://user-images.githubusercontent.com/121313267/211825284-f4030cc6-3ac3-4a0d-9740-751d405a26e2.jpg)
Sorry for the poor pictures.  Grabbed them quick on my way into work.  But that is the information from the BMS tool and the communication string.  I grabbed that because I wasn't sure if someone could compare to what the other EG4 communication looked like.

**Comment by pchiquit** [comment]
@Justice12t I have documented the EG4 24V protocol on this closed issue #212  . As you can see the EG4 24V uses the preamble `7e 01 01 00 fe 0d` compared to yours that seems to send `01 03 00 00 00 ..` , so the BMS seems to be different. I don't know how technical you are but the process of adding support for your battery is not that complex but would need some debugging from your part to see how close the protocol is from the EG4 24V.  This can be done by playing with the PC program and checking what commands are sent and received and comparing them to what is documented on my old issue.

**Comment by Justice12t** [comment]
That's what I was afraid of.  But ok fair.  I'm fairly technical and can usually trip through stuff.  I'll check out #212 and see where/if I can go from there.  Thanks for the help on this.

**Comment by Justice12t** [comment]
Would it be better if I just opened up a request to have this BMS added and the "battery" tag can be added to it?  I've found the attached document that looks like it breaks down the communications, I'm not sure if it is helpful to the right folks or not.  I did spend some time looking and know some of the responses.  I can add that information to the request if needed.  Let me know, thanks.
[work_Communication.txt](https://github.com/Louisvdw/dbus-serialbattery/files/10405979/work_Communication.txt)

**Comment by Justice12t** [comment]
@pchiquit @Louisvdw 
I attached my break down of my BMS for the EG4 12v/400.  Hopefully it makes sense.  I don't have all the answers but figured folks much smarter than me might be able to lend a hand.  I still have a few fields at the end to figure out which hopefully I'll get to this week.  I used the screenshot I posted above for the information.  Let me know if this is helpful or a complete waste of time.  Thanks!
 
[BMS Breakdown.txt](https://github.com/Louisvdw/dbus-serialbattery/files/10728387/BMS.Breakdown.txt)
[EG4-LL-MODBUS-Communication-Protocol_ENG-correct-1.pdf](https://github.com/Louisvdw/dbus-serialbattery/files/10728389/EG4-LL-MODBUS-Communication-Protocol_ENG-correct-1.pdf)

**Comment by pchiquit** [comment]
@Justice12t let me look and see if I can add this. I'll need a few days to find time to work on that. But this is exactly what I needed.

**Comment by Justice12t** [comment]
Awesome!  Thank you.  I've got a few minor updates.  The full capacity looks like it needs to be converted 1440000000 milliampere = 1440000 ampere, 1440000/3600 = 400 AH.  Also it starts with 01 03 4E... the 01 seems to correspond with the dip switch setting, so in this case it's set to 1 or battery 1.  Sure you already know this but just wanted to make sure I updated the info.  I'll see if I can work on the model, FW and Serial # this weekend.  Thanks again.  Let me know what else I can do to help.

**Comment by Justice12t** [comment]
@pchiquit 
I updated this file (at the bottom) with the serial #, FW version and Model.  I'm also including a file of the most recent communication detail that I used to figure this out.

[BMS Breakdown.txt](https://github.com/Louisvdw/dbus-serialbattery/files/10778233/BMS.Breakdown.txt)
[work_Communication.txt](https://github.com/Louisvdw/dbus-serialbattery/files/10778234/work_Communication.txt)

**Comment by Justice12t** [comment]
@pchiquit just wanted to check in and see how things were going and if there's anything else you needed from me.  Let me know, thanks!

**Comment by pchiquit** [comment]
@Justice12t my apologies. Life got busy with a vacation the wife wanted to take, so I was not able to start. Maybe next weekend I can make some progress.

**Comment by Justice12t** [comment]
@pchiquit no need to apologize. IF you aren't able to find the time no big deal either as I think I found another route so just let me know.  Thanks!

**Comment by Deemg003** [comment]
@pchiquit If you need another tester, let me know. 
Running 3x 12v400ah on Cerbo

**Comment by Justice12t** [comment]
@Louisvdw can this get "tagged" with battery?  I'm assuming that's how it gets added to the list to be reviewed and possibly code added? I believe we have all the pieces needed and I just didn't want it to fall off the list.  Thanks!

**Comment by tuxntoast** [comment]
@Justice12t  - I think we might have the same battery.  wrote something after finding some documentation on their version of ModBus

 https://github.com/tuxntoast/eg4-ll/blob/main/egll.py

**Reply by Justice12t** [reply]
Hey @tuxntoast just saw this haven't been on in a while.  Awesome, I'll check it out.  Stupid question, whats the best way to get it installed?  If I save it to a flash will it just boot? or do I need to do more?  Thanks.

**Reply by tuxntoast** [reply]
that file (egll.py) just needs to be present in the /data/etc/dbus-serialbatter/bms folder. It could get there via scp, or you could use vi, and paste the contents of the file into that editor. check the permissions on the file (ls -l), and set them to the correct user/permission with chmod / chown. You will need to add this driver to the /data/etc/dbus-serialbattery/dbus-serialbattery.py file. you do this via editing that file and insert this  "from bms.lifepower import egll"  to line 27 and add "{"bms": egll, "baud": 9600, "address":b"\x7C"}," insert this to line 44. After all this is done, you need to run the reinstall script via  /data/etc/dbus-serialbattery/reinstall-local.sh, then start your driver backup with "/data/etc/start-serial ttyUSB0" if the BMS is connected to usb0 on the cerbo. 

all the "code" is in quote above, remove those before adding that text to the dbus-serialbattery.py file. after you start the drive you can tail the log file to see the status of the battery and make sure all is well. that command is "tail -F -n 100 /data/log/dbus-serialbattery.ttyUSB0/current | tai64nlocal"

**Reply by Justice12t** [reply]
Thanks, I’ll give this a try this week.

**Reply by Justice12t** [reply]
I completely forgot about this.  Has this been added to the base code here?  or does it still need to be added per the instructions above?  Thanks.

**Reply by mr-manuel** [reply]
There was no PR to merge this. You are welcome to do so.

**Reply by Justice12t** [reply]
Is it ok if I request it? or should @tuxntoast do it since it is his code?

**Reply by mr-manuel** [reply]
See https://louisvdw.github.io/dbus-serialbattery/general/supported-bms#addrequest-new-bms

**Reply by KrispySmarTThings** [reply]
Hello all. I am in the beginning stages of installing my Cerbo and figured I would look to see if my EG4 12V 400A were supported. It seems that the solution exists in this topic.

Was this ever merged into the main software?

**Reply by tuxntoast** [reply]
Hey all! Sorry work has consumed me a little more then when I wrote this. 

I haven't done the PR yet. Let me look over what is required to get it into the official release.

**Reply by KrispySmarTThings** [reply]
All good! Quick question: When multiple of these batteries are in use, is it necessary to connect all of the batteries (3 in my case) to the Cerbo?- Or can I leave them "daisy chained" and just send the "master" to the Cerbo?

**Reply by tuxntoast** [reply]
I am pulling hardware serial numbers when I am connecting / reporting back stats, so the 3 usb trick should work. I think I had found a battery combiner that could sum those 3, and present a single battery to the cerbo.

As for battery that are chained together, and one cable to the cerbo, I really do not know how it will work. I only have a single EG4-LL 12v 400AH battery. I would assume there would be some refactor needed, but not sure what would be required until I could get access to that hardware setup.
